pkg/controller/clusterconnection/clusterconnection_controller.go

```diff
@@ -461,3 +466,45 @@ func (r *ReconcileConnection) Reconcile(ctx context.Context, request reconcile.R
 	// We should create the Guardian deployment.
 	return result, nil
 }
+
+// maintainInstallationFinalizer manages this controller's finalizer on the Installation resource.
+// We add a finalizer to the Installation when Guardian has been installed, and only remove that finalizer when
+// the Guardian has been deleted and its pods have stopped running. This allows for a graceful cleanup of Guardian resources
+// prior to the CNI plugin being removed.
+func maintainInstallationFinalizer(ctx context.Context, c client.Client, mgmtClusterConnection *operatorv1.ManagementClusterConnection) error {
+	// Get the Installation.
+	installation := &operatorv1.Installation{}
+	if err := c.Get(ctx, utils.DefaultInstanceKey, installation); err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.V(1).Info("Installation config not found")
+			return nil
+		}
+		log.Error(err, "An error occurred when querying the Installation resource")
+		return err
+	}
+	patchFrom := client.MergeFrom(installation.DeepCopy())
+
+	// Determine the correct finalizers to apply to the Installation. If the Whisker exists, we should apply
+	// a finalizer. Otherwise, if the Whisker namespace doesn't exist we should remove it. This ensures the finalizer
+	// is always present so long as the resources managed by this controller exist in the cluster.
+	if mgmtClusterConnection != nil {
+		// Add a finalizer indicating that the Goldmane is still running.
+		utils.SetInstallationFinalizer(installation, render.GuardianFinalizer)
+	} else {
+		// Check if the Guardian namespace exists, and remove the finalizer if not. Gating this on Namespace removal
+		// in the best way to approximate that all Guardian related resources have been removed.
+		l := &corev1.Namespace{}
+		err := c.Get(ctx, types.NamespacedName{Name: render.GuardianNamespace}, l)
```

Comment:
This isn't going to work because there is no Guardian Namespace - it runs within calico-system, which is going to exist until all of this teardown logic is completed. 

We need to check the Guardian deployment itself.